### DISCUSSION TITLE
NVIDIA-SMI

### DISCUSSION THREAD

**Discussion by alexander-lazarin** [gpuocelot/gpuocelot#13] [Q&A]
Is it possible to use gpuocelot in place of an actual NVIDIA GPU?
I've installed gpuocelot on my Ubuntu 22.04 but when I'm running `nvidia-smi` I get
`NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.`

**Comment by dmikushin** [comment]
Hi @alexander-lazarin , in our GPUOcelot test suite we run executables that depend on CUDA driver runtime and CUDA runtime this way:

```
LD_PRELOAD=/path/to/libgpuocelot.so nvidia-smi
```

Unfortunately, this method alone does not cover `nvidia-smi`, because this utility also touches low-level things, such as `/proc/driver/nvidia/params`. Emulating them should be easy though, but isn't done yet.

Please try to run the [deviceQuery sample](https://github.com/NVIDIA/cuda-samples/tree/master/Samples/1_Utilities/deviceQuery) first and let us know.

**Reply by dmikushin** [reply]
More precisely, `nvidia-smi` also wants to use `libnvidia-ml.so` NVIDIA Management Library (NVML), which communicates to the driver in a low-level way. In order to make it happy, GPUOcelot would have to re-create the device nodes it wants to touch.

**Reply by alexander-lazarin** [reply]
Hi @dmikushin, thank you for a quick response. I don't have any NVIDIA GPUs, so the result of deviceQuery is
```
./deviceQuery Starting...

 CUDA Device Query (Runtime API) version (CUDART static linking)

cudaGetDeviceCount returned 100
-> no CUDA-capable device is detected
Result = FAIL
``` 
My understanding is that using GPUOcelot as a simple replacement for a real NVIDIA device isn't possible at the time. Do you think this is something that might be developed in the future?

**Reply by dmikushin** [reply]
Just to double-check: have you preloaded the GPUOcelot library when executing the `deviceQuery` app?
